File: examples/PinnedH2O/rotation_test.py
import numpy as np

O1 = np.array([0.00, 0.00, 0.00])
H1 = np.array([-0.45, 1.42, -1.07])
H2 = np.array([-0.45, -1.48, -0.97])

def calculate_bondlength(atom1, atom2):
    return np.linalg.norm(atom1 - atom2)

def calculate_bondangle(atom1, atom2, atom3, radian):
    vector1 = atom1 - atom2
    vector2 = atom3 - atom2
    dot_product = np.dot(vector1, vector2)
    magnitude_product = np.linalg.norm(vector1) * np.linalg.norm(vector2)
    angle = np.arccos(dot_product / magnitude_product)
    if not radian:
        angle = np.degrees(angle)
    return angle

def rotation_matrix(axis, angle):
    cos_theta = np.cos(angle)
    sin_theta = np.sin(angle)
    ux, uy, uz = axis
    return np.array([
        [cos_theta + ux**2 * (1 - cos_theta), ux * uy * (1 - cos_theta) - uz * sin_theta, ux * uz * (1 - cos_theta) + uy * sin_theta],
        [uy * ux * (1 - cos_theta) + uz * sin_theta, cos_theta + uy**2 * (1 - cos_theta), uy * uz * (1 - cos_theta) - ux * sin_theta],
        [uz * ux * (1 - cos_theta) - uy * sin_theta, uz * uy * (1 - cos_theta) + ux * sin_theta, cos_theta + uz**2 * (1 - cos_theta)]
    ])

plane_normal = np.cross(H2, H1)
plane_normal = plane_normal / np.linalg.norm(plane_normal)

target_plane_normal = np.array([0, 0, 1])
axis_to_align = np.cross(plane_normal, target_plane_normal)
axis_to_align /= np.linalg.norm(axis_to_align)
angle_to_align = np.arccos(np.clip(np.dot(plane_normal, target_plane_normal), -1.0, 1.0))

rot_matrix_align_plane = rotation_matrix(axis_to_align, angle_to_align)
H1_rotated = np.dot(rot_matrix_align_plane, H1)
H2_rotated = np.dot(rot_matrix_align_plane, H2)

bondlength1 = calculate_bondlength(H1, O1)
bondlength2 = calculate_bondlength(H2, O1)
bondangle = calculate_bondangle(H1, O1, H2, False)

print('Original system')
print(f'H1 = {H1}')
print(f'H2 = {H2}')
print(f'Bondlength of O1-H1 = {bondlength1}')
print(f'Bondlength of O1-H2 = {bondlength2}')
print(f'Angle between O1-H1 and O1-H2 = {bondangle}')

bondlength1 = calculate_bondlength(H1_rotated, O1)
bondlength2 = calculate_bondlength(H2_rotated, O1)
bondangle = calculate_bondangle(H1_rotated, O1, H2_rotated, False)

print('Aligned system')
print(f'H1 = {H1_rotated}')
print(f'H2 = {H2_rotated}')
print(f'Bondlength of O1-H1 = {bondlength1}')
print(f'Bondlength of O1-H2 = {bondlength2}')
print(f'Angle between O1-H1 and O1-H2 = {bondangle}')

File: examples/PinnedH2O_3DOF/rotation_test.cc
#include <iostream>
#include <cmath>
#include <algorithm>
#include <cstring>

using namespace std;

double calculate_bondlength(const double atom1[3], const double atom2[3]) {
    return sqrt(pow(atom1[0] - atom2[0], 2) + pow(atom1[1] - atom2[1], 2) + pow(atom1[2] - atom2[2], 2));
}

double calculate_bondangle(const double atom1[3], const double atom2[3], const double atom3[3], bool radian) {
    double vector1[3] = {atom1[0] - atom2[0], atom1[1] - atom2[1], atom1[2] - atom2[2]};
    double vector2[3] = {atom3[0] - atom2[0], atom3[1] - atom2[1], atom3[2] - atom2[2]};

    double dot_product = vector1[0] * vector2[0] + vector1[1] * vector2[1] + vector1[2] * vector2[2];
    double magnitude_product = sqrt(pow(vector1[0], 2) + pow(vector1[1], 2) + pow(vector1[2], 2)) *
                               sqrt(pow(vector2[0], 2) + pow(vector2[1], 2) + pow(vector2[2], 2));
    double angle = acos(dot_product / magnitude_product);

    if (!radian) {
        angle = angle * 180.0 / M_PI;
    }
    return angle;
}

void rotation_matrix(const double axis[3], double angle, double matrix[3][3]) {
    double cos_theta = cos(angle);
    double sin_theta = sin(angle);
    double ux = axis[0], uy = axis[1], uz = axis[2];

    matrix[0][0] = cos_theta + ux * ux * (1 - cos_theta);
    matrix[0][1] = ux * uy * (1 - cos_theta) - uz * sin_theta;
    matrix[0][2] = ux * uz * (1 - cos_theta) + uy * sin_theta;

    matrix[1][0] = uy * ux * (1 - cos_theta) + uz * sin_theta;
    matrix[1][1] = cos_theta + uy * uy * (1 - cos_theta);
    matrix[1][2] = uy * uz * (1 - cos_theta) - ux * sin_theta;

    matrix[2][0] = uz * ux * (1 - cos_theta) - uy * sin_theta;
    matrix[2][1] = uz * uy * (1 - cos_theta) + ux * sin_theta;
    matrix[2][2] = cos_theta + uz * uz * (1 - cos_theta);
}

void normalize(double vec[3]) {
    double norm = sqrt(vec[0] * vec[0] + vec[1] * vec[1] + vec[2] * vec[2]);
    if (norm > 0) {
        vec[0] /= norm;
        vec[1] /= norm;
        vec[2] /= norm;
    }
}

void cross(const double a[3], const double b[3], double result[3]) {
    result[0] = a[1] * b[2] - a[2] * b[1];
    result[1] = a[2] * b[0] - a[0] * b[2];
    result[2] = a[0] * b[1] - a[1] * b[0];
}

void apply_rotation(const double matrix[3][3], const double vec[3], double result[3]) {
    result[0] = matrix[0][0] * vec[0] + matrix[0][1] * vec[1] + matrix[0][2] * vec[2];
    result[1] = matrix[1][0] * vec[0] + matrix[1][1] * vec[1] + matrix[1][2] * vec[2];
    result[2] = matrix[2][0] * vec[0] + matrix[2][1] * vec[1] + matrix[2][2] * vec[2];
}

int main() {
    double O1[3] = {0.00, 0.00, 0.00};
    double H1[3] = {-0.45, -1.48, -0.97};
    double H2[3] = {-0.45, 1.42, -1.07};

    double plane_normal[3];
    cross(H2, H1, plane_normal);
    normalize(plane_normal);

    double target_plane_normal[3] = {0, 0, 1};
    double axis_to_align[3];
    cross(plane_normal, target_plane_normal, axis_to_align);
    normalize(axis_to_align);
    double dot_product = plane_normal[0] * target_plane_normal[0] +
                         plane_normal[1] * target_plane_normal[1] +
                         plane_normal[2] * target_plane_normal[2];
    double angle_to_align = acos(min(max(dot_product, -1.0), 1.0));

    double rot_matrix_align_plane[3][3];
    rotation_matrix(axis_to_align, angle_to_align, rot_matrix_align_plane);

    double bondlength1 = calculate_bondlength(H1, O1);
    double bondlength2 = calculate_bondlength(H2, O1);
    double bondangle = calculate_bondangle(H1, O1, H2, false);

    cout << "Original system" << endl;
    cout << "H1 = (" << H1[0] << ", " << H1[1] << ", " << H1[2] << ")" << endl;
    cout << "H2 = (" << H2[0] << ", " << H2[1] << ", " << H2[2] << ")" << endl;
    cout << "Bondlength of O1-H1 = " << bondlength1 << endl;
    cout << "Bondlength of O1-H2 = " << bondlength2 << endl;
    cout << "Angle between O1-H1 and O1-H2 = " << bondangle << endl;

    double H1_rotated[3], H2_rotated[3];
    apply_rotation(rot_matrix_align_plane, H1, H1_rotated);
    apply_rotation(rot_matrix_align_plane, H2, H2_rotated);
    bool flipped_bond = false;

    if (bondlength1 < bondlength2) {
        flipped_bond = true;
        swap(H1_rotated, H2_rotated);
    }

    bondlength1 = calculate_bondlength(H1_rotated, O1);
    bondlength2 = calculate_bondlength(H2_rotated, O1);
    bondangle = calculate_bondangle(H1_rotated, O1, H2_rotated, false);

    cout << "Reference system (z=0 plane about x=0 axis, with longer bondlength in H1):" << endl;
    cout << "H1 = (" << H1_rotated[0] << ", " << H1_rotated[1] << ", " << H1_rotated[2] << ")" << endl;
    cout << "H2 = (" << H2_rotated[0] << ", " << H2_rotated[1] << ", " << H2_rotated[2] << ")" << endl;
    cout << "Bondlength of O1-H1 = " << bondlength1 << endl;
    cout << "Bondlength of O1-H2 = " << bondlength2 << endl;
    cout << "Angle between O1-H1 and O1-H2 = " << bondangle << endl;

    return 0;
}
